### PR TITLE
Cache external ontologies

### PR DESCRIPTION
**Pull request by wsobel**
Merged master and will update the About files.

### CODE REVIEW

**Comment by MilosDrobnjakovic** [comment]
I tested the about files, it properly loads via protege. Wasn't there also a plan to have an 'about' file in the Core directory as well?
Also, why are the Commons directories loaded as cache files in the Dev file but not the prod file? As far as I see it pulls the CMNS from the web in case the 'prod' file is loaded, while in the Dev it loads our cached version.

**Comment by wsobel** [comment]
Yes. It would be identical to the core file in the root. We can copy prod there. dev would be the same as prod in this case. 

We can also consider a file that load all the addenda swirl and property chains. 

Last, we can consider one that loads QUDT. 

These can be pushed. Please create the file(s) and add them to the PR. 

Thanks
- W

**Comment by MilosDrobnjakovic** [comment]
Do we need an about file for the supply chain ontology given that it is in the Dev phase?

**Comment by wsobel** [comment]
I would recommend. I don't know about the dependencies, so someone would need to provide.

**Comment by wsobel** [comment]
There should be NO Prod about files for Maintenance and SupplyChiain

**Comment by jlloganiii** [comment]
Right now the AboutIOFProd.rdf file works because https://spec.industrialontologies.org/iof/ontology/cache_external_ontologies/latest/cache/bfo/2020/bfo.rdf redirects to the "cache_external_ontologies" branch on the GitHub server. Before committing the merge, won't this have to change to "master", so it won't break if/when the  "cache_external_ontologies" branch is deleted? Same thing for all the lines in AboutIOFDev.rdf?

**Comment by wsobel** [comment]
It will need to be changed before the PR is merged.

File: AboutIOFDev.rdf
Comment (mereolog) [inline AboutIOFDev.rdf:32]: I would suggest using the actual IRI of the ontology here, i.e., `http://purl.obolibrary.org/obo/bfo.owl` and then "redirecting" it to the cache using an entry in https://github.com/iofoundry/ontology/blob/master/catalog-v001.xml.
Then software like Protege first will attempt resolve the IRI and if this fails, it will get it from the location specified in catalog-v001.xml. I.e., this additional entry may look like:

`<uri id="289c79d8-24a7-4ad6-a7d0-49da596ebe07" name="http://purl.obolibrary.org/obo/bfo.owl" uri="./cache/bfo/2020/bfo.rdf"/>`

Similar remarks apply to other import statements related to the "cached" ontologies.

**Review by mereolog — CHANGES_REQUESTED** [review]

**Comment by mereolog** [comment]
> Right now the AboutIOFProd.rdf file works because https://spec.industrialontologies.org/iof/ontology/cache_external_ontologies/latest/cache/bfo/2020/bfo.rdf redirects to the "cache_external_ontologies" branch on the GitHub server. Before committing the merge, won't this have to change to "master", so it won't break if/when the "cache_external_ontologies" branch is deleted? Same thing for all the lines in AboutIOFDev.rdf?

Actually, nothing is being redirected here - to see this run 
`curl -v https://spec.industrialontologies.org/iof/ontology/cache_external_ontologies/latest/cache/bfo/2020/bfo.rdf >/dev/null`
which should return a 200 response.

What happens is that our infrastructure keeps the content of cache_external_ontologies branch in the spec.industrialontologies.org server and simply resolves the contents of the respective file stored there.

But I think https://github.com/iofoundry/ontology/pull/77#discussion_r1507244203 makes this issue moot.

File: AboutIOFDev.rdf
Comment (wsobel) [reply AboutIOFDev.rdf:32]: We require that all external ontologies are stored in the cache directory. 

We will sync and maintain the ontologies.

Comment (ElisaKendall) [reply AboutIOFDev.rdf:32]: @mereolog There have been issues with BFO access, so the decision was made to cache all external ontologies for IOF to guarantee that the versions used for the IOF are those that have been validated. The validated ones are in the cache, so that's what the about file is describing in this case. This does add maintenance overhead, but it ensures that the ontologies all work for anyone using IOF.

**Review by wsobel — COMMENTED** [review]
We have made these changes and rewrote the IRIs to match what will be in the released cached URI.

File: AboutIOFDev.rdf
Comment (mereolog) [inline AboutIOFDev.rdf:32]: Given the current version of the catalog-v001.xml file this should read now http://purl.obolibrary.org/obo/bfo/2020/bfo.owl.

Similar remarks about other imports - you should use the ontology IRIs you added to the catalog-v001.xml file.

**Review by mereolog — CHANGES_REQUESTED** [review]

**Comment by MilosDrobnjakovic** [comment]
@wsobel the about files as it currently stands can't resolve BFO cached version - I assume that the IRI should become resolvable only when we merge to master. As for the individual ontologies, since its given in the catalog file, they properly load the cached ontologies. Any recommendations how to do the review? Should we have the 'proper resolution' in the catalog as well for good measure?

File: AboutIOFProd.rdf
Comment (MilosDrobnjakovic) [inline AboutIOFProd.rdf:22]: this is universal accross the About files  but I am going to log it just here: I recommend deleting the double Ontology declarations - the ontology is already declared on line 26

**Comment by wsobel** [comment]
Please make changes necessary to fix the problem and push.

**Comment by MilosDrobnjakovic** [comment]
> Please make changes necessary to fix the problem and push.

does this pertain to double ontology declaration, modfying the catalog file or both?

File: AboutIOFDev.rdf
Comment (jlloganiii) [reply AboutIOFDev.rdf:32]: The goal is to make loading an "about" file from an IRI correctly load BFO and the exact versions of Commons that are needed. Therefore, line 32 above should not change.

Comment (mereolog) [reply AboutIOFDev.rdf:32]: > The goal is to make loading an "about" file from an IRI correctly load BFO and the exact versions of Commons that are needed. Therefore, line 32 above should not change.

Well, I am not sure what you mean by "exact version", but as of now, the import statements in quesion will fail - see, for example, the snip below.

The reason for this is the difference betweeen the IRIs of the imported ontologies in the about file(s) and in the catalog-v001.xml file in [111744a](https://github.com/iofoundry/ontology/pull/77/commits/111744a8533af90e89edf04acb454d15d0899bb8) comit. So either you need to change the about file or the catalog-v001.xml.

My recommendation is based on the following idea of cache: you cache an ontology in case the IRI of this ontology will not resolve. So if it does resolve, the ontology is taken from the resolution site; otherwise, it is taken from the IOF cache folder.

image

Comment (wsobel) [reply AboutIOFDev.rdf:32]: I may have missed one of the corrections we were making yesterday. I'll check. 

If the loads for the catalog or the about files are not coming from cache, please fix them. 

The about files are now referencing the base IRI that should be available once it becomes the default version. We need, so the about files can only be tested once we merge and the cache path is no only in a branch. 

We will test once we have finished the merge and force push if necessary if the about files don't work.

Comment (wsobel) [reply AboutIOFDev.rdf:32]: The missing imports are because we are prepping for a release. The files _should_ be there once we have completed the merge and tagged.

File: AboutIOFProd.rdf
Comment (wsobel) [reply AboutIOFProd.rdf:22]: We need to wait for the merge.

File: AboutIOFDev.rdf
Comment (mereolog) [reply AboutIOFDev.rdf:32]: > The missing imports are because we are prepping for a release. The files _should_ be there once we have completed the merge and tagged.

Let me go back to https://github.com/iofoundry/ontology/pull/77#discussion_r1507244203.

Once you have 
<uri id="289c79d8-24a7-4ad6-a7d0-49da596ebe07" name="http://purl.obolibrary.org/obo/bfo.owl" uri="./cache/bfo/2020/bfo.rdf"/>
in the catalog file and
<owl:imports rdf:resource="http://purl.obolibrary.org/obo/bfo.owl"/>
everything you need will be in place right aways - you will not have to wait for the release.

I can see that for some reason you may want to freeze a particular version of bfo and use https://spec.industrialontologies.org/iof/ontology/cache/bfo/2020/bfo.rdf for this purpose, but this would require a change (back) in the catalog file and the change in cache/bfo/2020/bfo.rdf file itself so that the latter is identified by means of https://spec.industrialontologies.org/iof/ontology/cache/bfo/2020/bfo.rdf.

**Review by MilosDrobnjakovic — APPROVED** [review]
I have tested the working of the catalog files for maintenance, SCO, and Core. All properly load bfo from cache. Since for any other changes we are waiting for the merge I approve the request

**Comment by wsobel** [comment]
Thanks a lot. We will make changes to main if necessary once the merge completes.

**Comment by wsobel** [comment]
We still require a review.

**Review by jlloganiii — APPROVED** [review]
We can't fully test and fix problems until the merge is completed, so let's merge.

**Pull request merged**
